### PR TITLE
Lecture and assignment merge

### PR DESCRIPTION
**Pull request by dtracers**
mergers from #463 
Depends on #486, #487

### CODE REVIEW

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/Institution.java [hunk 1]
```diff
@@ -7,13 +7,12 @@
...
-import protobuf.srl.lecturedata.Lecturedata.Lecture;
 import protobuf.srl.lecturedata.Lecturedata.LectureSlide;
 import protobuf.srl.request.Message;
-import protobuf.srl.school.School.SrlAssignment;
-import protobuf.srl.school.School.SrlBankProblem;
+import protobuf.srl.school.Assignment.SrlAssignment;
+import protobuf.srl.school.Problem.SrlBankProblem;
 import protobuf.srl.school.School.SrlCourse;
-import protobuf.srl.school.School.SrlProblem;
+import protobuf.srl.school.Problem.SrlProblem;
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/Institution.java]: Move this a line higher so the problem are next to each other

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/AssignmentManager.java [hunk 2]
```diff
@@ -53,10 +50,56 @@
...
+ *     LECTURE, (this means it will look like a lecture)
+ *     GRADED, (this is the default students can not see any answers till the assignment is closed)
+ *     PRACTICE (this allows students to see the correct answer immediately)
+ *
+ * NavigationTypes:
+ *     DEFAULT, (chooses what the default type is for the specific assignmentType
+ *     LOOPING, (this means the question order loops when the end is reached)
+ *     NO_BACK_TRAVEL, (what it sounds like, as soon as they progress they can not navigate backwards [this one might require server navigation])
+ *     RANDOM (pulls the next one at a random order)
+ *
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/AssignmentManager.java]: What is all this?

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/BankProblemManager.java [hunk 3]
```diff
@@ -169,21 +159,15 @@ private static SrlBankProblem extractBankProblem(final DBObject dbObject, final
...
-            if (dbObject.get(BASE_SKETCH) != null) {
-                exactProblem.setBaseSketch(Commands.SrlUpdateList.parseFrom((byte[]) dbObject.get(BASE_SKETCH)));
+            if (dbObject.containsField(DatabaseStringConstants.SPECIAL_QUESTION_DATA)) {
+                exactProblem.setSpecialQuestionData(
+                        SlideManager.createElementFromQuery((DBObject) dbObject.get(DatabaseStringConstants.SPECIAL_QUESTION_DATA)));
             }
-        } catch (InvalidProtocolBufferException e) {
+        } catch (DatabaseAccessException e) {
+            LOG.error("Error parsing lecture element", e);
             e.printStackTrace();
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/BankProblemManager.java]: why print stack trace to console after you log it?

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/CourseProblemManager.java [hunk 4]
```diff
@@ -250,42 +277,311 @@ public static boolean mongoUpdateCourseProblem(final Authenticator authenticator
...
             update = true;
         }
         if (update) {
-            courseProblemCollection.update(cursor, new BasicDBObject(SET_COMMAND, updateObj));
+            final BasicDBObject setData = new BasicDBObject();
+            if (updateObj.size() > 0) {
+                setData.append(SET_COMMAND, updateObj);
+            }
+
+            // We only replace the list right now instead of adding new objects.  There are other methods for that.
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/CourseProblemManager.java]: Remove this block??

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/CourseProblemManager.java [hunk 5]
```diff
@@ -250,42 +277,311 @@ public static boolean mongoUpdateCourseProblem(final Authenticator authenticator
...
+
+        if (dbProblem.containsField(PROBLEM_NUMBER)) {
+            problem.setProblemNumber((Integer) dbProblem.get(PROBLEM_NUMBER));
+        }
+
+        return problem;
+    }
+
+    /**
      *
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/CourseProblemManager.java]: remove this line plz

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/CourseProblemManager.java [hunk 6]
```diff
@@ -250,42 +277,311 @@ public static boolean mongoUpdateCourseProblem(final Authenticator authenticator
...
+                createSlideProblemHolderQuery(slideId, Util.ItemType.SLIDE, unlocked));
+        courseProblemCollection.update(cursor, new BasicDBObject(ADD_SET_COMMAND, updateObj));
+
+        UserUpdateHandler.insertUpdates(dbs, ((List) cursor.get(USERS)), assignmentId, UserUpdateHandler.LECTURE_CLASSIFICATION);
+        return true;
+    }
+
+    /**
+     * NOTE: This is meant for internal use do not make this method public.
+     * <p/>
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/CourseProblemManager.java]: remove this tag

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/MongoInstitution.java [hunk 7]
```diff
@@ -26,17 +26,18 @@
...
-import protobuf.srl.school.School;
-import protobuf.srl.school.School.SrlAssignment;
-import protobuf.srl.school.School.SrlBankProblem;
+import protobuf.srl.school.Problem;
+import protobuf.srl.utils.Util;
+import protobuf.srl.school.Assignment.SrlAssignment;
+import protobuf.srl.school.Problem.SrlBankProblem;
 import protobuf.srl.school.School.SrlCourse;
-import protobuf.srl.school.School.SrlProblem;
+import protobuf.srl.school.Problem.SrlProblem;
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/MongoInstitution.java]: move this one line up

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/MongoInstitution.java [hunk 8]
```diff
@@ -426,8 +431,9 @@ public String insertBankProblem(final String userId, final String authId, final
     }
 
     @Override
-    public void updateLecture(final String authId, final Lecture lecture) throws AuthenticationException, DatabaseAccessException {
-        LectureManager.mongoUpdateLecture(auth, database, lecture.getId(), authId, lecture);
+    public void updateLecture(final String authId, final SrlAssignment lecture) throws AuthenticationException, DatabaseAccessException {
+        throw new DatabaseAccessException("lecture update not supported");
+        // LectureManager.mongoUpdateLecture(auth, database, lecture.getId(), authId, lecture);
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/MongoInstitution.java]: remove this

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/MongoInstitution.java [hunk 9]
```diff
@@ -456,7 +466,8 @@ public void updateBankProblem(final String authId, final SrlBankProblem srlBankP
 
     @Override
     public void updateLectureSlide(final String authId, final LectureSlide lectureSlide) throws AuthenticationException, DatabaseAccessException {
-        SlideManager.mongoUpdateLectureSlide(auth, database, lectureSlide.getId(), authId, lectureSlide);
+        throw new DatabaseAccessException("slide update not supported");
+        // SlideManager.mongoUpdateLectureSlide(auth, database, lectureSlide.getId(), authId, lectureSlide);
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/MongoInstitution.java]: remove this

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/MongoInstitution.java [hunk 10]
```diff
@@ -361,39 +362,43 @@ public String insertAssignment(final String userId, final String authId, final S
     }
 
     @Override
-    public String insertLecture(final String userId, final String authId, final Lecture lecture) throws AuthenticationException,
+    public String insertLecture(final String userId, final String authId, final SrlAssignment lecture) throws AuthenticationException,
             DatabaseAccessException {
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/MongoInstitution.java]: This method does not throw exceptions

File: CourseSketchProjects/CourseSketchDatabaseServer/src/test/java/database/institution/mongo/MongoInstitutionTest.java [hunk 11]
```diff
@@ -87,18 +90,18 @@
 
     public static final Commands.SrlUpdateList.Builder FAKE_UPDATELIST = Commands.SrlUpdateList.newBuilder();
     public static final Util.QuestionType FAKE_QUESTION_TYPE = Util.QuestionType.FREE_RESP;
-    public static final School.SrlAssignment.AssignmentType VALID_ASSIGNMENT_TYPE = School.SrlAssignment.AssignmentType.EXAM;
-    public static final int VALID_ASSIGNMENT_TYPE_VALUE = School.SrlAssignment.AssignmentType.EXAM_VALUE;
+    public static final Assignment.AssignmentType VALID_ASSIGNMENT_TYPE = Assignment.AssignmentType.GRADED;
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/test/java/database/institution/mongo/MongoInstitutionTest.java]: Isn't this Assignment.SrlAssignment rather then just Assignment

File: CourseSketchProjects/CourseSketchDatabaseServer/src/test/java/database/institution/mongo/MongoInstitutionTest.java [hunk 12]
```diff
@@ -87,18 +90,18 @@
...
+    public static final int VALID_ASSIGNMENT_TYPE_VALUE = Assignment.AssignmentType.GRADED_VALUE;
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/test/java/database/institution/mongo/MongoInstitutionTest.java]: Same here^

File: CourseSketchProjects/CourseSketchDatabaseServer/src/test/test stuff/TestProblemsSmall.java [hunk 13]
```diff
@@ -1,11 +1,11 @@
...
-import protobuf.srl.school.School.SrlAssignment.LatePolicy;
-import protobuf.srl.school.School.SrlBankProblem;
-import protobuf.srl.school.School.SrlBankProblem.QuestionType;
+import protobuf.srl.school.Assignment.SrlAssignment;
+import protobuf.srl.school.Assignment.SrlAssignment.LatePolicy;
+import protobuf.srl.school.Problem.SrlBankProblem;
+import protobuf.srl.school.Problem.SrlBankProblem.QuestionType;
 import protobuf.srl.utils.Util.SrlPermission;
-import protobuf.srl.school.School.SrlProblem;
+import protobuf.srl.school.Problem.SrlBankProblem;
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/test/test stuff/TestProblemsSmall.java]: You don't need this. look at line 5

File: CourseSketchProjects/coursesketchwebclient/src/main/src/common/navigation/problem/problemNavigator.js [hunk 14]
```diff
@@ -142,10 +149,17 @@ function ProblemNavigator(assignmentId, loop, preferredIndex) {
...
-        if (type !== CourseSketch.prutil.getSrlAssignmentClass().AssignmentType.GAME) {
+        var assignmentType = this.getAssignmentType();
+        var navigationType = this.getNavigationType();
+        if (isRandomNavigation(assignmentType, navigationType)) {
+            //Pull problems at random for Game
+            var numberOfQuestions = getProblemListSize();
+            var randomNumber = Math.random();
+
+            index = randomNumber % numberOfQuestions;
+        } else {
```
Comment (areftw) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/common/navigation/problem/problemNavigator.js]: Turn this into an else if {...}

File: CourseSketchProjects/coursesketchwebclient/src/main/src/common/navigation/problem/problemNavigator.js [hunk 15]
```diff
@@ -142,10 +149,17 @@ function ProblemNavigator(assignmentId, loop, preferredIndex) {
...
             if (index < 0 || index >= problemList.length && !loop) {
                 return;
             } else if (loop) {
```
Comment (areftw) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/common/navigation/problem/problemNavigator.js]: make:
`if (index >= problemList.length) {...}`
into:
`else if (index >= problemList.length) {...}`

**Comment by areftw** [comment]
Make sure this passes travis before merging

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/AssignmentManager.java (hunk 2 above)
Comment (dtracers) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/AssignmentManager.java]: documentation about what it stores

**Pull request merged**